File: simplerpg/Pixel.cpp
#include "Pixel.h"

extern const char *PixelColourNames[] = { "black", "blue", "green", "cyan", "red", "magenta", "yellow", "white" };

Pixel::Pixel(void)
{
	bold = false;
	mColour = 7;
	graphic = ' ';
}

Pixel::Pixel(char graphic, int colour, bool bold)
{
	this->graphic = graphic;
	this->bold = bold;
	setColour(colour);
}

Pixel::~Pixel(void)
{
}

void Pixel::savePixel(FormattedFile &file)
{
	const char *colourName = PixelColourNames[getColour()];
	file << colourName << ' ' << (bold ? '1' : '0') << ' ' << graphic << '\n';
}

void Pixel::loadPixel(FormattedFileIterator &iter)
{
	int colour = getColourFromString(*iter);	++iter;
	if (colour >= 0)
	{
		setColour(colour);
	}
	bold = (*iter)[0] == '1';					++iter;
	graphic = (*iter)[0];						++iter;
}

int Pixel::getColourFromString(const string &colourName)
{
	char c = colourName[0];
	if(c >= '0' && c <= '8')
	{
		return c - '0';
	}
	else
	{
		for(int i = 0; i < 8; i++)
		{
			if(iequals(colourName, PixelColourNames[i]))
			{
				return i;
			}
		}
		return -1;
	}
}

File: SimpleRPGAnalyser/Pixel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleRPGAnalyser
{
    public class Pixel : ILoadable
    {
        public string[] COLOUR_NAMES = { "black", "blue", "green", "cyan", "red", "magenta", "yellow", "white" };

        public int colour;
        public bool bold;
        public char graphic;

        public Pixel()
        {
            colour = 0;
            bold = false;
            graphic = 'x';
        }

        public int fullColour
        {
            get
            {
                if (bold)
                {
                    return colour | 0x8;
                }
                return colour;
            }
        }

        override public string ToString()
        {
            return graphic + "_" + colour.ToString() + '_' + (bold ? '1' : '0');
        }

        public void load(ref string[] iter, ref int index)
        {
            string c = iter[index++];
            if (c[0] >= '0' && c[0] < '9')
            {
                colour = int.Parse(c);
            }
            else
            {
                for (int i = 0; i < COLOUR_NAMES.Length; i++)
                {
                    if (c.Equals(COLOUR_NAMES[i], StringComparison.CurrentCultureIgnoreCase))
                    {
                        colour = i;
                        break;
                    }
                }
            }
            c = iter[index++];
            bold = c[0] == '1';
            graphic = iter[index++][0];
        }
    }
}
